src/ucm/bistro/bistro.c

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2018.       ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <dlfcn.h>
+#include <stdlib.h>
+
+#include <ucm/bistro/bistro.h>
+#include <ucm/bistro/bistro_int.h>
+
+#if HAVE___CLEAR_CACHE
+void __clear_cache(char* beg, char* end);
+#endif
+
+ucs_status_t ucm_bistro_remove_restore_point(ucm_bistro_restore_point_h rp)
+{
+    ucs_assert(rp != NULL);
+    free(rp);
+    return UCS_OK;
+}
+
+static void *ucm_bistro_page_align_ptr(void *ptr)
+{
+    return (void*)ucs_align_down((uintptr_t)ptr, ucm_get_page_size());
+}
+
+static ucs_status_t ucm_bistro_protect(void *addr, size_t len, int prot)
+{
+    void *aligned = ucm_bistro_page_align_ptr(addr);
+    size_t size   = addr - aligned + len;
+    int res;
+
+    res = mprotect(aligned, size, prot) ? UCS_ERR_INVALID_PARAM : UCS_OK;
+    if (res) {
+        ucm_error("Failed to change page protection: %m");
+        return UCS_ERR_INVALID_PARAM;
+    }
+    return UCS_OK;
+}
+
+ucs_status_t ucm_bistro_apply_patch(void *dst, void *patch, size_t len)
+{
+    ucs_status_t status;
+
+    status = ucm_bistro_protect(dst, len, UCM_PROT_READ_WRITE_EXEC);
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+
+    memcpy(dst, patch, len);
+
+    status = ucm_bistro_protect(dst, len, UCM_PROT_READ_EXEC);
+    if (!UCS_STATUS_IS_ERR(status)) {
+#if HAVE___CLEAR_CACHE
+        __clear_cache(dst, dst + len);
```

Comment:
What happens if you don't have it ?
Why it is needed ? It is the same thread, why do you have to flush cache ?

Reply:
on most modern CPUs there is separate cache for instructions and data and on weak memory coherency architectures self-modifying code needs to flush dcache and invalidate icache to

Reply:
I would expect clear cache to call ISB (checked llvm and it calls). not sure if we need this barrier.

Reply:
added usc_clear_cache call